evm/src/cpu/kernel/asm/secp256k1/lift_x.asm

```diff
@@ -0,0 +1,825 @@
+// Returns y such that (x,y) is on Secp256k1 and y&1 = v - 27,
+// as well as a flag indicating whether such a y exists.
+%macro secp_lift_x
+    // stack: x, v
+    %cubemodn_secp_base
+    // stack: x^3, v
+    PUSH 7
+    // stack: 7, x^3, v
+    %addmodn_secp_base
+    // stack: x^3+7, x, v
+    DUP1
+    // stack: x^3+7, x^3+7, v
+    %sqrt_secp_base
+    // stack: y, x^3+7, x, v
+    SWAP1
+    // stack: x^3+7, y, v
+    DUP2
+    // stack: y, x^3+7, y, v
+    %squaremodn_secp_base
+    // stack: y^2, x^3+7, y, v
+    EQ
+    // stack: sqrtOk, y, v
+    SWAP2
+    // stack: v, y, sqrtOk
+    DUP2
+    // stack: y, v, y, sqrtOk
+    PUSH 1
+    // stack: 1, y, v, y, sqrtOk
+    AND
+    // stack: 1 & y, v, y, sqrtOk
+    PUSH 27
+    // stack: 27, 1 & y, v, y, sqrtOk
+    SWAP1
+    // stack: 1 & y, 27, v, y, sqrtOk
+    SWAP2
+    // stack: v, 27, 1 & y, y, sqrtOk
+    SUB
+    // stack: v - 27, 1 & y, y, sqrtOk
+    EQ
+    // stack: correctParity, y, sqrtOk
+    DUP2
+    // stack: y, correctParity, y, sqrtOk
+    %secp_base
+    // stack: N, y, correctParity, y, sqrtOk
+    SUB
+    // stack: N - y, correctParity, y, sqrtOk
+    SWAP1
+    // stack: correctParity, N - y, y, sqrtOk
+    %select_bool
+    // stack: goody, sqrtOk
+%endmacro
+
+%macro cubemodn_secp_base
+    // stack: x
+    DUP1
+    // stack: x, x
+    %squaremodn_secp_base
+    // stack: x^2, x
+    %mulmodn_secp_base
+%endmacro
+
+%macro addmodn_secp_base
+    // stack: x, y
+    %secp_base
+    // stack: N, x, y
+    SWAP2
+    // stack: y, x, N
+    ADDMOD
+%endmacro
+
+// Returns sqrt(x). Computed as x^(q+1)/4, with q the Secp base field order.
+/// To replace with more efficient method using non-determinism later.
+%macro sqrt_secp_base
+    // stack: x
+    DUP1
+    %squaremodn_secp_base
+    DUP2
```

Comment:
I just made #613 to add a `%rep` syntax, if it looks good to you maybe you could hit merge and use it here

Reply:
So I generate these exponentiation functions with a small Python script:
```python
for b in reversed(exponent.bits()[:-1]):
    print("%squaremodn")
    if b:
        print("DUP2\n%mulmodn")
```
Using `%rep` would make this script slightly more complicated and shouldn't change much since these functions should go away once we have non-determinism. Is there another reason to use `%rep` that I'm missing?

Reply:
Ok - we can keep it as is for now